src/core/symbology/classification/qgsclassificationjenks.cpp

```diff
@@ -0,0 +1,171 @@
+/***************************************************************************
+    qgsclassificationjenks.h
+    ---------------------
+    begin                : September 2019
+    copyright            : (C) 2019 by Denis Rouzaud
+    email                : denis@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <limits> // for jenks classification
+#include "qgsclassificationjenks.h"
+
+QgsClassificationJenks::QgsClassificationJenks()
+  : QgsClassificationMethod( true /*valuesRequired*/,
+                             false /*symmetricMode*/ )
+{
+
+}
+
+QString QgsClassificationJenks::name() const
+{
+  return QObject::tr( "Natural Breaks (Jenks)" );
+}
+
+QString QgsClassificationJenks::id() const
+{
+  return QStringLiteral( "Jenks" );
+}
+
+QgsClassificationMethod *QgsClassificationJenks::clone() const
+{
+  QgsClassificationJenks *c = new QgsClassificationJenks();
+  copyBase( c );
+  return c;
+}
+
+
+QList<double> QgsClassificationJenks::calculateBreaks( double minimum, double maximum,
+    const QList<double> &values, int nclasses )
+{
+  // Jenks Optimal (Natural Breaks) algorithm
+  // Based on the Jenks algorithm from the 'classInt' package available for
+  // the R statistical prgramming language, and from Python code from here:
+  // http://danieljlewis.org/2010/06/07/jenks-natural-breaks-algorithm-in-python/
+  // and is based on a JAVA and Fortran code available here:
+  // https://stat.ethz.ch/pipermail/r-sig-geo/2006-March/000811.html
+
+  // Returns class breaks such that classes are internally homogeneous while
+  // assuring heterogeneity among classes.
+
+  if ( values.isEmpty() )
+    return QList<double>();
+
+  if ( nclasses <= 1 )
+  {
+    return QList<double>() <<  maximum;
+  }
+
+  if ( nclasses >= values.size() )
+  {
+    return values;
+  }
+
+  QVector<double> sample;
+
+  // if we have lots of values, we need to take a random sample
+  if ( values.size() > mMaximumSize )
+  {
+    // for now, sample at least maximumSize values or a 10% sample, whichever
```

Comment:
is it worth moving the support for max sample sizes outside of this, so that the feature request used to fetch the values has a setLimit on it?

Reply:
That makes sense but a beyond the scope of this PR. I believe this can be easily added later on, even in classification base class.